packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -383,20 +385,45 @@ function useAutoFocus(scopeRef: RefObject<HTMLElement[]>, autoFocus: boolean) {
       }
     }
     autoFocusRef.current = false;
-  }, []);
+  }, [scopeRef]);
 }
 
-function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boolean, contain: boolean) {
+// An array to create a stack of nodeToRestore elements.
+let nodeToRestoreArray: HTMLElement[] = [];
+
+function addToNodeToRestoreArray(node: HTMLElement) {
+  if (nodeToRestoreArray.indexOf(node) === -1) {
+    nodeToRestoreArray.push(node);
+  }
+}
+
+function removeFromNodeToRestoreArray(node: HTMLElement) {
+  let index = nodeToRestoreArray.indexOf(node);
+  if (index > -1) {
+    nodeToRestoreArray.splice(index, 1);
+  }
+}
+
+function useRestoreFocus(scopeRef: RefObject<HTMLElement[]>, restoreFocus: boolean | RefObject<DOMRefValue<HTMLElement>> | RefObject<HTMLElement>, contain: boolean) {
   // create a ref during render instead of useLayoutEffect so the active element is saved before a child with autoFocus=true mounts.
   const nodeToRestoreRef = useRef(typeof document !== 'undefined' ? document.activeElement as HTMLElement : null);
 
   // useLayoutEffect instead of useEffect so the active element is saved synchronously instead of asynchronously.
   useLayoutEffect(() => {
-    let nodeToRestore = nodeToRestoreRef.current;
     if (!restoreFocus) {
       return;
     }
 
+    let nodeToRestore: HTMLElement;
+
+    if (typeof restoreFocus === 'boolean') {
+      nodeToRestore = nodeToRestoreRef.current as HTMLElement;
+    } else if (restoreFocus.current) {
+      nodeToRestore = restoreFocus.current instanceof HTMLElement ? restoreFocus.current : restoreFocus.current.UNSAFE_getDOMNode();
+    }
+
+    addToNodeToRestoreArray(nodeToRestore);
+
```

Comment:
IMO we should still save the `document.activeElement` in the `nodeToRestoreArray` even if the user provides a node to `restoreFocus` to cover for a case where the supplied `restoreFocus` node is removed before the FocusScope is unmounted.

Reply:
That make sense.